src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyStatus.cs

```diff
@@ -0,0 +1,299 @@
+            }
+        }
+
+        internal bool CanAdjustConcurrency()
+        {
+            // Don't adjust too often, either up or down if we've made an adjustment recently.
+            TimeSpan timeSinceLastAdjustment = DateTime.UtcNow - LastConcurrencyAdjustmentTimestamp;
+            int minAdjustmentFrequencyMs = GetLatencyAdjustedInterval(1000 * ProcessMonitor.DefaultSampleIntervalSeconds * 2, DefaultMinAdjustmentFrequencySeconds * 1000, 1);
+
+            return timeSinceLastAdjustment.TotalMilliseconds > minAdjustmentFrequencyMs;
+        }
+
+        internal bool CanDecreaseConcurrency()
+        {
+            return CurrentConcurrency > 1;
+        }
+
+        internal bool CanIncreaseConcurrency(int maxDegreeOfParallelism)
+        {
+            var timeSinceLastDecrease = DateTime.UtcNow - LastConcurrencyDecreaseTimestamp;
+            int minDecreaseQuietWindowMs = GetLatencyAdjustedInterval(1000 * ProcessMonitor.DefaultSampleIntervalSeconds * 10, DefaultFailedAdjustmentQuietWindowSeconds * 1000, 10);
+            if (timeSinceLastDecrease != null && timeSinceLastDecrease.TotalMilliseconds < minDecreaseQuietWindowMs)
+            {
+                // if we've had a recent failed adjustment, we'll avoid any increases for a while
+                return false;
+            }
+
+            if (MaxConcurrentExecutionsSinceLastAdjustment < CurrentConcurrency)
+            {
+                // We only want to increase if we're fully utilizing our current concurrency level.
+                // E.g. if we increased to a high concurrency level, then events slowed to a trickle,
+                // we wouldn't want to keep increasing.
+                return false;
+            }
+
+            // a max parallelism of -1 indicates unbounded
+            return maxDegreeOfParallelism == -1 || CurrentConcurrency < maxDegreeOfParallelism;
+        }
+
+        internal int GetLatencyAdjustedInterval(int minIntervalMS, int defaultIntervalMS, int latencyMultiplier)
+        {
+            int resultInterval = defaultIntervalMS;
+
+            if (InvocationsSinceLastAdjustment > 0)
+            {
+                // Compute based on function latency, so faster functions can adjust more often, allowing their concurrency
+                // to "break away" from longer running, heavier functions.
+                // While a longer running function might not actually be a problem (e.g. might be I/O bound),
+                // this latency based prioritization is still beneficial. It allows fast functions to give
+                // us fast feedback. Worst case, for long running functions we just revert back to the default
+                // interval. So this is an optimization applied when possible.
+                int avgInvocationLatencyMs = (int)TotalInvocationTimeSinceLastAdjustmentMs / InvocationsSinceLastAdjustment;
+                int computedInterval = minIntervalMS + latencyMultiplier * avgInvocationLatencyMs;
```

Comment:
Open to ideas on a better formula for this.

Reply:
It's just a way to scale the latency factor used in the calculation to adjust the result.